File: graph/TopologicalSort/cpp/TopologicalSort.cpp
#include "TopologicalSort.hpp"

#include <queue>
#include <vector>

using namespace std;

vector<int> TopologicalSort(const vector<vector<int>>& graph) {
    vector<int> indegree(graph.size(), 0);
    for (const auto& neighbors : graph) {
        for (const int neighbor : neighbors) {
            ++indegree[neighbor];
        }
    }

    queue<int> ready;
    for (int node = 0; node < static_cast<int>(graph.size()); ++node) {
        if (indegree[node] == 0) {
            ready.push(node);
        }
    }

    vector<int> order;
    while (!ready.empty()) {
        const int node = ready.front();
        ready.pop();
        order.push_back(node);

        for (const int neighbor : graph[node]) {
            --indegree[neighbor];
            if (indegree[neighbor] == 0) {
                ready.push(neighbor);
            }
        }
    }

    return order;
}

File: graph/TopologicalSort/java/TopologicalSort.java
import java.util.*;

public class TopologicalSort {
    public static List<Integer> TopologicalSort(List<List<Integer>> graph) {
        int[] indegree = new int[graph.size()];
        for (List<Integer> neighbors : graph) {
            for (int neighbor : neighbors) {
                indegree[neighbor]++;
            }
        }

        Deque<Integer> ready = new ArrayDeque<>();
        for (int node = 0; node < graph.size(); node++) {
            if (indegree[node] == 0) {
                ready.add(node);
            }
        }

        List<Integer> order = new ArrayList<>();
        while (!ready.isEmpty()) {
            int node = ready.remove();
            order.add(node);

            for (int neighbor : graph.get(node)) {
                indegree[neighbor]--;
                if (indegree[neighbor] == 0) {
                    ready.add(neighbor);
                }
            }
        }

        return order;
    }
}
